src/main/java/org/apache/sling/resourcemerger/impl/MergingResourceProvider.java

```diff
@@ -51,86 +51,55 @@
         this.traverseHierarchie = traverseHierarchie;
     }
 
-    protected static final class ExcludeEntry {
-
-        public final String name;
-        public final boolean exclude;
-        public final boolean onlyUnderlying; // if only underlying resources should be affected (and not the local ones)
-
-        public ExcludeEntry(final String value, boolean onlyUnderlying) {
-            this.onlyUnderlying = onlyUnderlying;
-            if ( value.startsWith("!!") ) {
-                this.name = value.substring(1);
-                this.exclude = false;
-            } else if ( value.startsWith("!") ) {
-                this.name = value.substring(1);
-                this.exclude = true;
-            } else {
-                this.name = value;
-                this.exclude = false;
-            }
-        }
-    }
-
     /**
      * Class to check whether a child resource must be hidden. It should not be instantiated for the underlying resource
      * tree (which is /libs by default) because this check is expensive.
      */
-    protected static final class ParentHidingHandler {
+    protected static final class ResourceHidingHandler {
 
-        private List<ExcludeEntry> entries = new ArrayList<ExcludeEntry>();
+        private boolean isParentHiddenFully;
+        private boolean isParentHiddenForUnderlay;
+        private final HideItemPredicate hidePredicate;
 
         /**
          *
-         * @param parent the underlying resource
+         * @param resource the underlying resource
          * @param traverseParent if true will also continue with the parent's parent recursively
          */
-        public ParentHidingHandler(final Resource parent, final boolean traverseParent) {
+        public ResourceHidingHandler(final Resource resource, final boolean traverseParent) {
+            isParentHiddenFully = false;
+            isParentHiddenForUnderlay = false;
             // evaluate the sling:hideChildren property on the current resource
-            final ValueMap parentProps = parent.getValueMap();
-            final String[] childrenToHideArray = parentProps.get(MergedResourceConstants.PN_HIDE_CHILDREN, String[].class);
+            final ValueMap properties = resource.getValueMap();
+            final String[] childrenToHideArray = properties.get(MergedResourceConstants.PN_HIDE_CHILDREN, String[].class);
```

Comment:
If you call `properties.get(MergedResourceConstants.PN_HIDE_CHILDREN, new String[0])` instead, you can eliminate the following null check and the one for `hidePredicate != null` further down.

Reply:
Fixed in https://github.com/apache/sling-org-apache-sling-resourcemerger/pull/4/commits/32cf5745f8dce1246209396d631a96445d49adbb